src/librustc_typeck/check/mod.rs

```diff
@@ -4376,7 +4376,7 @@ pub fn check_bounds_are_used<'tcx>(tcx: TyCtxt<'tcx>, generics: &ty::Generics, t
         } else if let ty::Error = leaf_ty.sty {
             // If there is already another error, do not emit
             // an error for not using a type Parameter.
-            assert!(tcx.sess.err_count() > 0);
+            assert!(tcx.sess.has_errors());
```

Comment:
Code like this should probably be using `delay_span_bug` uniformly.

Reply:
I'd prefer an `EmittedError` type only created by emitting errors and putting that in `ty::Error` instead.

Reply:
That's even nicer - we might want to use that in the AST and HIR as well.

But maybe it should be something less temporarily ordered, so if there is some place where `ty::Error` needs to be created but it's not the same place the errors is to be reported from, we can still create a proof that "some error has happened or will happen" with `delay_span_bug`.